crates/store/README.md

```diff
@@ -0,0 +1,829 @@
+
+```rust
+Iter<Structured<ContextState>, Structured<StateData>>
+```
+
+This type ensures that the iterator will yield keys of type `ContextState` and
+values of type `StateData`, providing compile-time guarantees about the types of
+data being processed.
+
+#### Integration with layers and transactions
+
+The iterator system is deeply integrated with other components of the library:
+
+  1. **Layer integration**: The `ReadLayer` trait includes an iter method that
+     returns an `Iter`. This allows each layer to provide its own implementation
+     of iteration, potentially adding additional functionality or optimisations.
+
+  2. **Transaction integration**: The `Transaction` struct provides methods for
+     creating iterators over its operations. This allows for efficient
+     processing of pending changes in a transaction.
+
+#### Range queries
+
+The iterator system supports efficient range queries through the use of a
+`start` key in the `iter()` method of `ReadLayer`. This allows for iteration to
+begin at a specific point in the key space, enabling efficient partial scans of
+the data.
+
+#### Performance considerations
+
+The iterator system is designed with performance in mind:
+
+  1. **Lazy evaluation**: Iterators are lazy, meaning they only do work when
+     asked for the next item. This allows for efficient processing of large
+     datasets without unnecessary memory usage.
+
+  2. **Zero-cost abstractions**: The use of Rust's trait system and generic
+     programming allows many of the abstractions to be resolved at compile-time,
+     resulting in efficient machine code.
+
+  3. **Custom implementations**: The `DBIter` trait allows for database-specific
+     optimisations in the underlying iteration logic.
+
+#### Extensibility
+
+The iterator system is designed to be extensible:
+
+  1. **Custom iterators**: New iterator types can be easily created by
+     implementing the `DBIter` trait.
+
+  2. **Composition**: The `IterPair` struct allows for the composition of
+     multiple iterators, enabling complex iteration patterns.
+
+#### Use cases
+
+The iterator system supports a wide range of use cases:
+
+  1. **Scanning**: Efficiently scan through large portions of the data store.
+
+  2. **Filtering**: Iterate over data and filter based on key or value
+     properties.
+
+  3. **Aggregation**: Perform aggregations or reductions over ranges of data.
+
+  4. **Prefix queries**: Efficiently query all keys with a certain prefix.
+
+#### Limitations and considerations
+
+While powerful, users of the iterator system should be aware of certain
+considerations:
+
+  1. **Consistency**: In a multi-threaded environment, iterators provide a
+     point-in-time view and may not reflect concurrent modifications. If an
+     iterator is to be shared across threads, care should be taken to ensure
+     consistency — and it may be preferable to use [Rayon][]'s parallel
+     iterators instead for concurrent processing.
```

Comment:
ah, it's mentioned here.. I think it needs better placement than down here

Also, iterators should behave no differently from other Rust types across threads, we don't need to mention Rayon.

Rust won't let incorrect use compile.

Reply:
I've added a link 👍

Reply:
> Also, iterators should behave no differently from other Rust types across threads, we don't need to mention Rayon.

That's... not quite right. It's true that in Rust, iterators themselves don't have special thread-safety properties. The behaviour of iterators across threads is determined by the types they iterate over and the ownership model of Rust. If a type is `Send` and `Sync`, it can be safely transferred and accessed across threads. This applies to iterators as well.

So although you are correct in implying that iterators don't inherently need special mention of Rayon regarding thread safety, as long as the underlying types conform to Rust's concurrency guarantees, it does not therefore automatically follow that everything iterable will work exactly as expected. Hence the mention of Rayon as a _possible_ ("it _may_ be preferable") option for concurrent processing.

Reply:
To clarify, I meant "our" iterators. In Rust, iterators take a `&mut self` which according to Rust's rules, forbids multiple references, that's where Rayon helps.

I mention this because you said "consistency", which I'm taking to mean "in the face of concurrent writes", your "point-in-time" view should stay "consistent" at that point in time.

In Rust, this is forbidden because the iterator you hold references some held data, while the mutation you intend to make, requires an exclusive reference.

In our case, `Database::{put, delete}` take shared references, implying that each database must implement snapshotting and manage RW contention internally.

This means we're able to provide an uncomplicated way to iterate without requiring caveat communication.

And as such, however you choose to consume the output iterator should be safe.

This is why I said, our iterators should behave no differently from other Rust types across threads.

The same considerations you would make for a `Vec`'s or `BTreeMap`'s iterator, is the same you would make here.

Which is why I said, it's not necessary to mention.

Reply:
Thanks for the clarification. You're right that Rust's type system prevents many incorrect uses at compile time, but when dealing with persistent storage and database-like structures there are additional considerations beyond what Rust's type system can enforce.

Even if the Rust code is safe, the data being iterated over could change if there are concurrent modifications to the database. This "consistency" is what I was referring to, rather than memory safety.

Separately, the mention of Rayon was indeed not necessary, but I felt it was a useful note for anyone needing to perform concurrent iteration. It is not critical, and I agree it can be removed.

You're absolutely right that our iterators should behave similarly to other Rust types in terms of thread safety, and I haven't disagreed with that, but equally the discussion there is going off at a tangent unrelated to what I wrote.

As the way I've structured the sentence has caused conflation, and the mention of Rayon is not necessary, I've simply removed it.